File: frontend/src/platform/types/role.ts
import type { PermissionOut } from './permission'

export interface RoleBase {
  name: string
  description?: string | null
}

export type RoleIn = RoleBase

export interface RolePatch {
  name?: string
  description?: string | null
}

export interface RolePermissionIn {
  permission_ids: number[]
}

export interface RoleOut extends RoleBase {
  id: number
  organization_id: number
  is_protected: boolean
  permissions: PermissionOut[]
  created_at: string
  updated_at: string
}

File: backend/src/platform/schemas/role.py
from typing import Annotated

from pydantic import AfterValidator, BaseModel, ConfigDict, Field

from src.platform.schemas.common import NameDescriptionOut, Timestamp
from src.platform.schemas.utils import sort_by_id


class RoleBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    description: str | None = None


class RoleOut(RoleBase, Timestamp):
    id: int
    organization_id: int
    is_protected: bool = False
    permissions: Annotated[list[NameDescriptionOut], AfterValidator(sort_by_id)] = (
        Field(default_factory=list)
    )


class RoleIn(RoleBase): ...


class RolePatch(BaseModel):
    name: str | None = None
    description: str | None = None


class RolePermissionIn(BaseModel):
    permission_ids: list[int] = Field(default_factory=list)
